[generator] Merge world motorways into trunks

pastk opened the pull request:
It'll allow for more efficient lines merging.

pastk commented:
@vng what was the issue with merging e.g. `highway=trunk_link` into `highway=trunk`?
```
    /// @todo It's not obvious to integrate link->way conversion.
    /// Review it in future.
```
https://github.com/organicmaps/organicmaps/commit/a22150b215d7a88ac67cd6f48335249751583773

vng commented:
I don't remember. Probably, SetMappingTypes doesn't work properly at all.

Long features are also not good. Should be some reasonable tradeoff here, but I can't formalize it.

pastk commented:
> I don't remember. Probably, SetMappingTypes doesn't work properly at all.

It worked in my small test generation. A motorway and a trunk were merged into one line.

> 
> Long features are also not good. Should be some reasonable tradeoff here, but I can't formalize it.

Should be possible to limit by e.g. X points or Y length, maybe the lines merger does it already? Though I'm not sure.

pastk commented:
> Long features are also not good. Should be some reasonable tradeoff here, but I can't formalize it.

World line lengths (by number of points):
```
average  9.5
median   22
max      5095
```

Zurich:
```
average  8.5
median   15
max      581
```

pastk commented:
This was included into the 231013 data generation in #6295 .
Leads to ~45Kb data savings.

It fixes some "gaps" also; before:
![image](https://github.com/organicmaps/organicmaps/assets/18434508/0cd0a659-93e9-40f7-bf94-fafb62589ab4)

after:
![image](https://github.com/organicmaps/organicmaps/assets/18434508/38026388-3d47-40fd-88b0-65b85d079703)


But surprisingly it also led to a appearance of new "gaps" e.g. here in the center:
![image](https://github.com/organicmaps/organicmaps/assets/18434508/218f46da-7dfa-458d-84f6-f2259dd33cb6)

So results are mixed overall.

pastk commented:
The short lines filtering algo and / or lines merging algo should be improved further.

The short line filtering is very dump, it just removes them without checking if it'll create a gap between other lines.
It kind of assumes all short lines will be merged into longer ones when possible, but it doesn't happen always.

E.g. in this case the merge algo will merge paraller parts of a double carriage-ways roads (red and blue) into single continuous lines (it doesn't take into account they're oneway and in opposite directions), so there will be a short connecting green line left inbetween and it'll be discarded by the short lines filter thus creating a "gap".
![image](https://github.com/organicmaps/organicmaps/assets/18434508/0a176a79-25eb-4e23-aa65-3574e0554e19)

pastk commented:
I've disabled discarding of too short lines for the World map (left it for ferry lines only).
The only other lines we have are boundaries and roads.
IMHO it doesn't make sense to filter roads like this, at least not with the current approach which produces a lot of "gaps".

Example of improvements:

![world-roads-LaRioja](https://github.com/organicmaps/organicmaps/assets/18434508/cf595add-c68a-4e03-b924-b6871267588c)

For the ferry lines, on the opposite, I made the "too short line" criteria more strict to remove shorter lines across rivers and narrower lakes.
E.g. in the following screenshot a short ferry line in the lower part of the lake had been discarded.

![world-roads-Montenegro](https://github.com/organicmaps/organicmaps/assets/18434508/9e2029f5-4644-4e15-a681-0fdc072e4d5d)

I've changed the merging logic to prefer to merge shorter lines first. It should help to spread the same amount of points between the line features more evenly by avoiding too long and too short lines.

It'd be good to include this change into the next data generation to test it more thoroughly.

pastk commented:
@vng @biodranik

vng commented:
Make sense. One more gaps source is junctions. They should be replaced with a direct connection in World map.

pastk commented:
Before/after comparison
![world-roads-z9_2](https://github.com/organicmaps/organicmaps/assets/18434508/0d2c7f6c-cdd3-4d68-bfbc-78d27c8277fe)

pastk commented:
Although there are apparent improvements (see a screenshot above) to roads continuity, the test world map size increased a lot (by 20Mb almost):
- `boundary-administrative-*` increased by several times due to #6445 
  - I'll re-enable the short lines filter for boundaries
- roads size increased 1.5-2 times which is too much for just filling in the gaps
- meta-data and search index sizes increased noticeably which is very strange and might mean we store some data for lines unnecessarily

biodranik commented:
> roads size increased 1.5-2 times which is too much for just filling in the gaps

That means that the simplification is not working properly _after_ joining segments.

pastk commented:
> * roads size increased 1.5-2 times which is too much for just filling in the gaps

Looks like the main contributor to size inflation are V-like approaches/links to roundabouts and other roads:
![shot_2023-11-08_10-46-13](https://github.com/organicmaps/organicmaps/assets/18434508/fa5c621d-4a82-4bbf-84d7-dfdc9b17c1ef)
Apparently only one side of V could be merged with the main road line and the other one is left as a very short separate line which has a big storage overhead.

E.g. a `Span_La Rioja` region has many of those Vs and its part of World map is inflated by almost 50%, while Macedonia doesn't have those mapped and its size is increased by 3% only.

So an interim solution here is to retain the short lines filter, but make the threshold more strict (discard only very-very short lines).

pastk commented:
- adjusted short lines filtering thresholds separately for boundaries, ferries and roads
- removed metadata from lines (~3% mwm size reduction)
- reduced points precision for World lines merger to be same as in lines simplificator (might help to merge more lines)

Ready for another test World generation.

pastk commented:
- removed metadata from areas also

pastk commented on generator/world_map_generator.hpp:
```diff
@@ -69,8 +91,12 @@ class WorldMapGenerator
+      // Discard name, metadata, etc. for lines and areas.
+      // TODO(pastk): filter points metadata, retain necessary bits only.
+      if (fb.GetGeomType() != feature::GeomType::Point)
+        fb.ResetParams();
```
Why not? Why keep unnecessary data?
(I don't expect any major savings of course)

vng replied:
- I don't like the way it is made
- I suspect we should store meta for area at least in future

pastk replied:
1. what is the problem exactly? before my change the metadata removal was "hidden" in the lines specific feature merger code, now its in a more logical and visible place
2. the we can easily change it if needed (and we should keep only **used** metadata, not **all** metadata)

vng commented on generator/world_map_generator.hpp:
```diff
@@ -53,9 +53,31 @@ class WorldMapGenerator
+      if (scales::IsGoodForLevel(thresholdLevel, fb.GetLimitRect()))
+      {
+        auto mfb = fb; // Could be modified in PushSure().
+        PushSure(mfb);
```
I don't like this.

pastk replied:
me too :) any idea how to make it better?

pastk commented on generator/feature_merger.cpp:
```diff
@@ -109,14 +109,16 @@ size_t MergedFeatureBuilder::GetKeyPointsCount() const
   for (size_t i = 1; i < poly.size(); ++i)
     pr += poly[i-1].SquaredLength(poly[i]);
-  return pr;
+  return std::numeric_limits<double>::max() - pr;
```
fixed

pastk commented:
@vng PTAL

biodranik commented on generator/world_map_generator.hpp:
```diff
@@ -100,6 +119,12 @@ class WorldMapGenerator
 
+    // Merge motorways into trunks.
+    // TODO : merge e.g. highway-trunk_link into highway-trunk?
+    char const * marr1[2] = {"highway", "motorway"},
```
constexpr

biodranik replied:
https://medium.com/@samathy_barratt/constantly-confusing-c-const-and-constexpr-pointer-behaviour-7eddb041e15a

biodranik commented on generator/world_map_generator.hpp:
```diff
@@ -53,8 +53,27 @@ class WorldMapGenerator
     {
-      // Skip small ways. This check is enough, because classifier types check was made in m_typesCorrector.
-      if (scales::IsGoodForLevel(scales::GetUpperWorldScale(), fb.GetLimitRect()))
+      static const uint32_t ferryType = classif().GetTypeByPath({"route", "ferry"});
```
constexpr

pastk replied:
GetTypeByPath() is not constexpr

biodranik replied:
Right! It would be great to make these expressions constexpr in the future...

biodranik reviewed (commented):
Did you test it?

pastk commented:
> Did you test it?

Too many times :) There are screenshots above.

But it needs to go through a test planet generation anyway.

vng commented:
I think I know where the useless search index comes from. Will investigate later.
